privacyidea/lib/policy.py

```diff
@@ -2563,6 +2600,7 @@ def token(cls, g, scope, action, token_obj):
         :param token_obj: The token where the user object or the realm should match.
         :rtype: ``Match``
         """
+        # Todo: eventually consider token serial given by g.serial
```

Comment:
We should discuss this shortly. Maybe we need additional tests?

Reply:
We will do. 
THanks a lot so far.

Reply:
A testcase:

policy "Always_deny_access" 
* Scope: authorization
* priority: 100
* action: authorize=deny_access

policy "Allow softtoken"
* Scope: authorization
* priority: 1
* action: authorize=grant_access
* extended condition: tokeninfo: tokenkind=software

# Expected Outcome

Software-Tokens can authenticate
Hardware-Tokens can not authenticate.

# Actual Outcome

Both tokens can authenticate, matching policy is "Allow softtoken".

# Reason

The tokeninfo is never evaluated, since the policy "authorize" uses the ``Match.user`` matcher, which does not pass the serial to the lower level. Only ``Match.generic`` does.

I have this in my setup and I can show you this tomorrow.

Reply:
One more thing, you will realize, when we actually use the tokeninfo condition.

The request must fail, if tokeninfo is not available - just like with the userinfo. You will realize it, when I show this tomorrow ;-)

Reply:
Added api tests and fixed lib test (error message has changed) in https://github.com/privacyidea/privacyidea/pull/2549/commits/35c99cc107cf220f254f5977b306e78803a79001

Reply:
> Wait a second. I think we actually can and need to do the following:
> 
> For all classmethods (realm, user, admin, admin_or_user, action_only) we need to do this following:
> 
> In the final object-constructor we need to pass the **serial** number, otherwise in nearly all policies the tokeninfo will not work out, since the serial is only check for policies that are verified for Match-objects with the `generic` constructor.
> 
> ```python
> return cls(g, name=None, scope=scope, realm=realm, active=True,
>                    resolver=None, user=None, user_object=None,
>                    client=g.client_ip, action=action, adminrealm=None, time=None, serial=g.serial,
>                    sort_by_priority=True)
> ```

class constructors now have a g.serial (https://github.com/privacyidea/privacyidea/pull/2549/commits/35c99cc107cf220f254f5977b306e78803a79001). No problems experienced so far.

Reply:
Note, that an active tokeninfo policy condition produces an exception whenever a wrong password is entered at /validate/check. The api test specifically covers that case.